apm-agent-core/src/main/java/co/elastic/apm/agent/configuration/ActivationType.java

```diff
@@ -0,0 +1,134 @@
+                case REMOTE: return APM_AGENT_ATTACH_CLI;
+                case SELF_ATTACH: return PROGRAMMATIC_SELF_ATTACH;
+                case FLEET: return FLEET;
+            }
+        }
+
+        //The next picks up `Container image' lambdas but only if the layer is used, which we
+        //aren't sure is good or bad. We could also check ELASTIC_APM_AWS_LAMBDA_HANDLER
+        //but have deliberately decided not to (for now)
+        String lambdaWrapper = PrivilegedActionUtils.getEnv("AWS_LAMBDA_EXEC_WRAPPER");
+        logger.debug("AWS_LAMBDA_EXEC_WRAPPER: {}", lambdaWrapper);
+        if (null != lambdaWrapper && lambdaWrapper.equals("/opt/elastic-apm-handler")) {
+            return AWS_LAMBDA_LAYER;
+        }
+
+        String elasticJavaagentOnTheCommandline = getElasticJavaagentOnTheCommandline();
+        logger.debug("Agent on command line: {}", elasticJavaagentOnTheCommandline);
+        String javaToolOptions = PrivilegedActionUtils.getEnv("JAVA_TOOL_OPTIONS");
+        logger.debug("JAVA_TOOL_OPTIONS: {}", javaToolOptions);
+        if (null != javaToolOptions) {
+            // fallback k8s test, this is the pattern used by the webhook, but it's
+            // a fragile check as the webhook could easily be changed
+            if (javaToolOptions.contains("-javaagent:/elastic/apm/agent/elastic-apm-agent.jar")) {
+                return K8S_ATTACH;
+            } else if(elasticJavaagentOnTheCommandline != null && javaToolOptions.contains(elasticJavaagentOnTheCommandline)) {
+                return ENV_ATTACH;
+            }
+        }
+
+        if(elasticJavaagentOnTheCommandline != null) {
+            return JAVAAGENT_FLAG;
+        }
+
+        return UNKNOWN;
+    }
+
+    public static String getAgentJarFilename() {
+        String agentLocation = PrivilegedActionUtils.getProtectionDomain(GlobalTracer.class).getCodeSource().getLocation().getFile();
+        if (agentLocation != null) {
+            String agentJarFile = agentLocation.replace('\\', '/');
+            if (agentJarFile.contains("/")) {
+                return agentJarFile.substring(agentLocation.lastIndexOf('/') + 1, agentLocation.length());
+            } else {
+                return agentJarFile;
+            }
+        } else {
+            return null;
+        }
+    }
+
+    public static String getElasticJavaagentOnTheCommandline() {
+        String agentJarFile = getAgentJarFilename();
+        if (agentJarFile != null) {
+            List<String> javaArgs = ManagementFactory.getRuntimeMXBean().getInputArguments();
+            if (javaArgs != null) {
+                //if there is more than one, this will return the first, which is the correct algorithm
+                for (String javaArg : javaArgs) {
+                    if (javaArg.startsWith("-javaagent:") && javaArg.contains(agentJarFile)) {
+                        return javaArg;
+                    }
+                }
+            }
+        }
+        return null;
+    }
```

Comment:
Since this basically finds out whether the `-javaagent` was used, can we just rely on whether the agent was started through `co.elastic.apm.agent.premain.AgentMain#premain()` or `co.elastic.apm.agent.premain.AgentMain#agentmain()`? Would we miss anything?
Note that this information already permeates to `co.elastic.apm.agent.bci.ElasticApmAgent#initialize`, so if this is good enough, it should be much easier to get.